clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.consumer.internals;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public enum MemberState {
+
+    /**
+     * Member has not joined a consumer group yet
+     */
+    UNJOINED,
+
+    /**
+     * Member has received a new assignment (partitions assigned or revoked), and it is applying
+     * it. While in this state, the member will invoke the user callbacks for
+     * onPartitionsAssigned or onPartitionsRevoked, and then make the new assignment effective.
+     */
+    // TODO: determine if separate state will be needed for assign/revoke (not for now)
+    PROCESSING_ASSIGNMENT,
```

Comment:
I guess - this is where we send the first heartbeat?

Reply:
This is when the member receives a new assignment in the hearbeat response. It stays in thit state as long as it is processing the assignment (onPartitionsAssigned and onPartitionsRevoked callbacks execution + make the assignment received the `currentAssignment`). 
This is the interesting state we've been talking a lot about, mainly because we will continue to receive heartbeat responses in this state, and we need to be able to process them correctly while processing the assignment.